### ISSUE TITLE
metabase only foreign key error when adding column to visualization

### ISSUE THREAD

**Issue by wollerman** [metabase/metabase#10279]
**Describe the bug**
Specifying a foreign key constraint within the metabase admin panel without the underlying databsase foreign key prevents adding columns under the `visualization` section by selecting more columns.

One quirk that I'm not sure is affecting this is one of my "tables" is a non-materialized view.

**Logs**
```
Jul 01 13:33:55 WARN metabase.query-processor.middleware.process-userland-query :: Query failure {:status :failed,
:class clojure.lang.ExceptionInfo,
:error "Value does not match schema: {:query {:fields (named (not (\"distinct\" a-clojure.lang.PersistentVector)) \"Distinct, non-empty sequence of Field clauses\")}}",
:stacktrace
("--> util.schema$schema_core_validator$fn__15674.invoke(schema.clj:26)"
"query_processor.middleware.validate$validate_query$fn__41919.invoke(validate.clj:9)"
"query_processor.middleware.normalize_query$normalize$fn__36125.invoke(normalize_query.clj:11)"
"query_processor.middleware.catch_exceptions$catch_exceptions$fn__35143.invoke(catch_exceptions.clj:104)"
"query_processor.middleware.process_userland_query$process_userland_query$fn__38649.invoke(process_userland_query.clj:161)"
"query_processor.middleware.constraints$add_default_userland_constraints$fn__35192.invoke(constraints.clj:42)"
"query_processor.middleware.async$async_setup$fn__34317.invoke(async.clj:106)"
"query_processor$fn__42059$process_query__42064$fn__42065.invoke(query_processor.clj:290)"
"query_processor$fn__42059$process_query__42064.invoke(query_processor.clj:284)"
"query_processor$fn__42083$process_query_and_save_execution_BANG___42088$fn__42089.invoke(query_processor.clj:312)"
"query_processor$fn__42083$process_query_and_save_execution_BANG___42088.invoke(query_processor.clj:306)"
"query_processor$fn__42107$process_query_and_save_with_max_results_constraints_BANG___42112$fn__42113.invoke(query_processor.clj:323)"
"query_processor$fn__42107$process_query_and_save_with_max_results_constraints_BANG___42112.invoke(query_processor.clj:317)"
"query_processor.async$fn__42180$process_query_and_save_with_max_results_constraints_BANG___42185$fn__42186.invoke(async.clj:37)"
"query_processor.async$fn__42180$process_query_and_save_with_max_results_constraints_BANG___42185.invoke(async.clj:32)"
"api.dataset$fn__45627.invokeStatic(dataset.clj:52)"
"api.dataset$fn__45627.invoke(dataset.clj:41)"
"middleware.auth$enforce_authentication$fn__61677.invoke(auth.clj:14)"
"routes$fn__62814$fn__62815.doInvoke(routes.clj:56)"
"middleware.exceptions$catch_uncaught_exceptions$fn__61768.invoke(exceptions.clj:104)"
"middleware.exceptions$catch_api_exceptions$fn__61765.invoke(exceptions.clj:92)"
"middleware.log$log_api_call$fn__63188$fn__63189.invoke(log.clj:170)"
"middleware.log$log_api_call$fn__63188.invoke(log.clj:164)"
"middleware.security$add_security_headers$fn__61731.invoke(security.clj:105)"
"middleware.json$wrap_json_body$fn__62893.invoke(json.clj:61)"
"middleware.json$wrap_streamed_json_response$fn__62911.invoke(json.clj:97)"
"middleware.session$bind_current_user$fn__60345$fn__60346.invoke(session.clj:193)"
"middleware.session$do_with_current_user.invokeStatic(session.clj:176)"
"middleware.session$do_with_current_user.invoke(session.clj:170)"
"middleware.session$bind_current_user$fn__60345.invoke(session.clj:192)"
"middleware.session$wrap_current_user_id$fn__60334.invoke(session.clj:161)"
"middleware.session$wrap_session_id$fn__60319.invoke(session.clj:123)"
"middleware.auth$wrap_api_key$fn__61685.invoke(auth.clj:27)"
"middleware.misc$maybe_set_site_url$fn__63253.invoke(misc.clj:58)"
"middleware.misc$bind_user_locale$fn__63256.invoke(misc.clj:74)"
"middleware.misc$add_content_type$fn__63241.invoke(misc.clj:30)"
"middleware.misc$disable_streaming_buffering$fn__63278.invoke(misc.clj:119)"
"middleware.misc$wrap_gzip$fn__63271.invoke(misc.clj:101)"),
:query
{:query
{:source-table 100,
:fields [["field-id" 1709] ["field-id" 1707] ["field-id" 1710] ["field-id" 1708] ["fk->" 1709 723] ["fk->" ["field-id" 1709] ["field-id" 723]]],
:filter ["not-null" ["field-id" 1709]]},
:type "query",
:parameters [],
:async? true,
:middleware {:add-default-userland-constraints? true, :userland-query? true},
:info
{:executed-by 1,
:context :ad-hoc,
:card-id nil,
:nested? false,
:query-hash [-28, -21, 63, -38, 46, 85, 91, 106, -4, 95, -14, -23, -87, 68, 26, 1, 29, -110, 114, -126, 106, -33, -82, 33, 27, 95, -110, -87, 96, -85, -48, 65]},
:constraints {:max-results 10000, :max-results-bare-rows 2000}},
:preprocessed nil,
:native nil,
:ex-data
{:type :schema.core/error,
:value
{:query
{:source-table 100,
:fields [[:field-id 1709] [:field-id 1707] [:field-id 1710] [:field-id 1708] [:fk-> [:field-id 1709] [:field-id 723]] [:fk-> [:field-id 1709] [:field-id 723]]],
:filter [:not-null [:field-id 1709]]},
:type :query,
:database 3,
:async? true,
:middleware {:add-default-userland-constraints? true, :userland-query? true},
:info
{:executed-by 1,
:context :ad-hoc,
:nested? false,
:query-hash [-28, -21, 63, -38, 46, 85, 91, 106, -4, 95, -14, -23, -87, 68, 26, 1, 29, -110, 114, -126, 106, -33, -82, 33, 27, 95, -110, -87, 96, -85, -48, 65]},
:constraints {:max-results 10000, :max-results-bare-rows 2000}},
:error {:query {:fields (named (not ("distinct" a-clojure.lang.PersistentVector)) "Distinct, non-empty sequence of Field clauses")}}}}
```

**To Reproduce**
Pre req for behaviour:
1. Create table A with some column `foo`
2. Create table B with some column `foo` and some column `bar`
3. Do not create a foreign key between them at the database level
4. Create a foreign key relationship in metabase via the admin `Data Model`
    a. e.g. specify table `A` column `foo` has type `Foreign Key` of `B -> foo`.

1. Go to `ask a question > custom`
2. Select `db > table A`
3. Click `Get Answer`
4. click the "gear" icon and try to select `bar` from table `B` under the `More columns` section.
5. See error

**Expected behavior**
The expectation is that these tables can be joined to see columns from both tables under the circumstances that `foo` is distinct and non-empty.

**Screenshots**
Effectively the screenshot just gives the user the same error as in the logs:
```
Value does not match schema: 
{:query {:fields (named (not ("distinct" a-clojure.lang.PersistentVector)) "Distinct, non-empty sequence of Field clauses")}}
```

**Information about your Metabase Installation:**
- Your browser and the version: Chrome 74.0.3729.169

- Your operating system: OS X

- Your databases: postgres 10

- Metabase version: v032.9

- Metabase hosting environment: aws

- Metabase internal database: postgres

**Severity**
For now it's just annoying. But to help give BI tools to our non-technical users this will cause frustration and difficulty. We can work around it by writing the query directly, but that defeats the purpose and elegance of using the metabase UI.

**Additional context**
I am able to run these queries with raw sql. So there is some constraint that metabase is enforcing to prevent this result. Specifically I've run the query:

```
select B.bar, A.* from A inner join B where A.foo = B.foo;
```

Honestly I'm not 100% sure what constraint is exactly being violated. Is it related to the query being generated? Or the underlying relationship that metabase expects? Or the fact that one of my "tables" is actually a view? Or something else entirely?

**Comment by flamber** [comment]
Hi @wollerman 
I'm almost certain that you're seeing that error because you're adding the field multiple times - and you're doing that because of a javascript problem (with no error in the console) prevented columns from being added. It's issue #9144 - and there's a fix for it in PR #10242 (already fixed on 0.33)
The [upcoming](https://discourse.metabase.com/t/preview-upcoming-changes-to-the-gui-query-builder-updated/6551) 0.33 changes a lot, so it's much easier to join tables in the editor.

**Comment by wollerman** [comment]
@flamber thanks for the quick reply! Great to hear that it's already addressed; I didn't realize it was the same issue. I'll close this and try out 0.33 when it's out of preview. If that doesn't solve it I'll follow up on #9144 or this issue as appropriate. Thanks again!
